kotlin-extension/src/main/kotlin/io/opentelemetry/example/kotlinextension/CoroutineContextExample.kt

```diff
@@ -0,0 +1,79 @@
+package io.opentelemetry.example.kotlinextension
+
+import io.opentelemetry.api.trace.Tracer
+import io.opentelemetry.context.Context
+import io.opentelemetry.extension.kotlin.asContextElement
+import io.opentelemetry.sdk.testing.exporter.InMemorySpanExporter
+import io.opentelemetry.sdk.trace.SdkTracerProvider
+import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.util.concurrent.CountDownLatch
+
+/**
+ * This example aims to show how to embed a Span context within a Kotlin coroutine so that any new span created inside the
+ * coroutine will automatically get the Span context provided for the coroutine even if said new span is created after
+ * changing the coroutine's thread.
+ */
+class CoroutineContextExample {
+    private val inMemoryExporter: InMemorySpanExporter = InMemorySpanExporter.create()
+    private val tracer by lazy { createTracer() }
+
+    fun run(countDownLatch: CountDownLatch) {
+        val rootSpan = tracer.spanBuilder("Root span").startSpan()
+
+        val rootScope = rootSpan.makeCurrent()
+        println("Root span trace ID: ${rootSpan.spanContext.traceId}")
+
+        // Create the coroutine scope to launch it later.
+        val scope = CoroutineScope(Dispatchers.IO)
+
+        // Launching a coroutine with the Span context embedded in its [the coroutine's] context. The `asContextElement` is a Kotlin
+        // extension function that ensures that the Span context will be preserved across the life of the coroutine
+        // regardless of any "thread change" (coroutine context change) that might happen inside the coroutine.
+        // We could also look at the line below as if we'd call `span.makeCurrent` but for a Kotlin coroutine. So after
+        // launching the coroutine this way (with the embedded Span context) any new span created inside the coroutine
+        // must automatically get our root span as its parent.
+
+        // Bonus: If we wanted to launch a coroutine in a separate thread (not IO, which is the one set in the scope) while
+        // keeping the Span context too, we could do so like this:
+        // scope.launch(Dispatchers.Main + Context.current().asContextElement())
+        scope.launch(Context.current().asContextElement()) {
+            // Creating spans that must have the rootSpan as their parent since its context is attached to the
+            // coroutine context.
+            createChildrenSpans()
+
+            rootScope.close()
+            rootSpan.end()
+
+            val finishedSpanItems = inMemoryExporter.finishedSpanItems
+            println("First child's trace id: ${finishedSpanItems[1].traceId}") // The same as the root span's.
+            println("Second child's trace id: ${finishedSpanItems[2].traceId}") // The same as the root span's.
```

Comment:
You would have a restructure a little bit, but what about moving this to the main method after the latch is awaited? That feels a bit more "finalized" to me and also more clearly shows that this is the demonstrated example output.

Reply:
Can you also add the `thread.id` as an attribute to the spans and print it here?  I guess there's no simple way to force the execution to be on a different thread, but it would be a nice visual verification that it works.

Reply:
Thanks, I've made some changes to it to clean it up a bit and to move the verification part outside of the coroutine block. I also like the idea of printing the `thread.id`, so that's added now as well.